### PR TITLE
Basic ATSU, AIDS, CFDS functionality + options menu

### PR DESCRIPTION
**Pull request by wpine215**
## Summary of Changes

- [X] [MCDU] Added new menu page and structure (credit: Edwin B)
- [X] [MCDU] Updated text on MCDU ident page to reflect proper engine name (LEAP-1A26)
- [X] [ATSU] Implemented AOC page (credit: Edwin B)
- [X] [ATSU] Implemented message viewing (credit: Edwin B)
- [X] [ATSU] Implemented METAR requests (credit: Edwin B)
- [x] [ATSU] Implemented D-ATIS requests
- [x] [ATSU] Implemented TAF requests
- [X] [ATSU] Implemented free text (telex) message sending
- [X] [ATSU] Implemented free text (telex) message receiving
- [X] [ATSU] Implemented COMPANY MSG upper right ECAM memo
- [X] [AIDS] Implemented placeholder page
- [X] [CFDS] Implemented some placeholder pages
- [x] [CFDS] Implement EIS/DMC display test modes
- [X] [OPTIONS] Implemented option for ADIRS alignment time
- [X] [OPTIONS] Implemented option for AOC METAR uplink source
- [x] [OPTIONS] Implemented option for AOC D-ATIS uplink source
- [x] [OPTIONS] Implemented option for AOC TAF uplink source
- [x] [OPTIONS] Implement option for DMC display self-test time
- [x] [OPTIONS] Implemented option for telex messaging (should be off by default)
- [X] [SERVER] Implemented backend API for METAR
- [X] [SERVER] Implemented backend API for D-ATIS
- [x] [SERVER] Implemented backend API for TAF
- [X] [SERVER] Implemented backend API for telex connections
- [X] [SERVER] Implemented backend API for telex messages
- [x] [SERVER] Separate API into multiple files
- [x] [SERVER] Use tokens rather than IP address for authentication
- [x] [MISC] Refactor code into separate functions (messages and HTTP requests)
- [x] [MISC] Add custom getter/setter for JS local storage

New features/fixes to add (as of 10/24):

Features to be included in upcoming separate PR:
- [ATSU] Add Simbrief flightplan import to INIT DATA REQUEST
- [MCDU] Fix font spacing issues
- [MCDU] Fix top right page numbering/arrows
- [MCDU] Add NAV B/UP option on top right (main menu)
- [OPTIONS] Add setting for Simbrief pilot ID
- [OPTIONS] Add settings for cabin/passenger/copilot sounds

Features moved to backlog:
- [ATSU] Implement SIGMET uplink (frontend + backend)
- [ATSU] Implement wind data uplink (frontend + backend)
- [CFDS] Add reverse thrust test
- [CFDS] Implement GPWC display and sound test mode
- [CFDS] Implement DCDU/ATIMS display test mode
- [CFDS] Implement TCAS test
- [MCDU] Add option to set CHG CODE (main menu)

## Screenshots (if necessary)

vlc_LCE6NwMPVR

Video:
https://youtu.be/CFfMkuhNl8Y

## References

To-do

https://www.youtube.com/watch?v=twwUjHXdNVU (Honeywell Pegasus 2)

## Additional context

Co-authored by Edwin B (ATSU)

Discord username (if different from GitHub): Iceman#5540

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build.py will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, slick on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by veikkos** [comment]
Might be worth mentionin in the description that where the data is from? Awesome feature.

**Comment by wpine215** [comment]
> 
> 
> Might be worth mentionin in the description that where the data is from? Awesome feature.

Data is pulled from MSFS's METAR API which I datamined from the game files (no in-game weather SDK functions at the moment). Routed through a Google Cloud API because direct requests from a locally hosted file are prevented because of CORS rules that MSFS set on their server.

**Comment by waredrive** [comment]
Are there any plans to make the METAR source configurable? I'm currently using "Unreal Weather" addon which  uses data from AVWX. The user needs to create an AVWX account and API token so it can be used when calling AVWX API. This is great as live METAR is used to create the weather in game. Will it be possible to implement smilar feature and use (real) live METAR information in MCDU?

**Comment by wpine215** [comment]
> 
> 
> Are there any plans to make the METAR source configurable? I'm currently using "Unreal Weather" addon which uses data from AVWX. The user needs to create an AVWX account and API token so it can be used when calling AVWX API. This is great as live METAR is used to create the weather in game. Will it be possible to implement smilar feature and use (real) live METAR information in MCDU?

I'm planning to add an options menu in the MCDU that allows you to set the METAR source. Currently it only pulls from MS servers, but I plan to add support for VATSIM, PilotEdge, and IVAO servers. I haven't considered AVWX yet, but I'll add it to the list for sure.

**Comment by waredrive** [comment]
> > Are there any plans to make the METAR source configurable? I'm currently using "Unreal Weather" addon which uses data from AVWX. The user needs to create an AVWX account and API token so it can be used when calling AVWX API. This is great as live METAR is used to create the weather in game. Will it be possible to implement smilar feature and use (real) live METAR information in MCDU?
> 
> I'm planning to add an options menu in the MCDU that allows you to set the METAR source. Currently it only pulls from MS servers, but I plan to add support for VATSIM, PilotEdge, and IVAO servers. I haven't considered AVWX yet, but I'll add it to the list for sure.

That sounds great, thank you!

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AocRequestsWeather.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            store["sendStatus"] = "QUEUED"
+            updateView();
+            const ICAOS = [store["arpt1"], store["arpt2"], store["arpt3"], store["arpt4"]];
+            mcdu.clearUserInput();
+            const lines = []; // Prev Messages
+            let errors = 0;
+            const getData = async () => {
+                for (const icao of ICAOS) {
+                    if (icao !== "") {
+                        await fetch(`https://us-central1-flybywire-metar.cloudfunctions.net/metar-py?source=ms&icao=${icao}`)
```
Comment (devsnek) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_AocRequestsWeather.js:74]: You could speed this up in the case of multiple requests by doing something like:
```js
const parts = await Promise.all(ICAOS
  .filter((icao) => icao !== '')
  .map(async (icao) => {
    const data = await fetch('...').then((r) => r.text());
    if (data.slice(0, 9) === 'FBW_ERROR') {
      return [
        // ...
      ];
    }
    return [
      `METAR ${icao}`,
      ...wordWrapToStringList(data, 25),
      '------------------------',
    ];
  });
const lines = parts.flat();

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_FreeText.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                updateView();
+                const recipient = store["msg_to"]
+                const msgLines = [store["msg_line1"], store["msg_line2"], store["msg_line3"], store["msg_line4"]].join(";");
+                mcdu.clearUserInput();
+                let errors = 0;
+                let endpoint = "https://api.flybywiresim.com/txmsg";
+
+                const getData = async () => {
+                    if (recipient != "" && msgLines != ";;;") {
+                        await fetch(`${endpoint}?from=${telexFlight}&to=${recipient}&message=${msgLines}`, {method: "POST"})
```
Comment (2hwk) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_FreeText.js:95]: I hope there is sanitisation of user input at the server end, i.e. preventing special characters, maximum message length, etc. else people will spam arabic unicode and webdings copypastas. Don't think this is open to XSS.

Comment (wpine215) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_FreeText.js:95]: Yep, I'll implement a server-side regex filter tomorrow to only allow alphanumeric characters, periods, and forward slashes, as well as a filter to replace slurs with asterisks.

**Comment by myselfaaryan** [comment]
https://youtu.be/77g8NzBCXzA

You can use this for reference 😄

**Comment by VCC18DT-Hornet** [comment]
intresting

**Comment by ImenesFBW** [comment]
This is an awesome PR.

First small bug I found is the MAINTENANCE MODE screen doesn't have the bluish background. (This was with start on the runway)

image

**Comment by VCC18DT-Hornet** [comment]
I am not too totally sure about this (Could be other factors causing this FPS hit) but I am seeing an FPS hit with this PR that I didn't have with the Latest Dev Version

**Comment by Benjozork** [comment]
@VCC18DT-Hornet

Can you film a video with the MSFS FPS indicator showing ?

**Comment by VCC18DT-Hornet** [comment]
> @VCC18DT-Hornet
> 
> Can you film a video with the MSFS FPS indicator showing ?

not really

**Comment by myselfaaryan** [comment]
I am no QA, but I did test this PR and I had no issues. I am pretty sure that my fps degradation is only with the seperate nd and pfd. Since they were turned off, I was getting my normal fps 30-35.

**Comment by Carter633** [comment]
Name: British Avgeek YT -- Carter633 #1436
Date of testing: Current date 29/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1280 
Object of test: Check out the new basic ATSU, AIDS, CFDS functionality

Testing technique: Remained on ground and fiddled (with the MCDU)
Steps you've made to test the Pull Request:
1. Spawned C&D, ext power and Battery on
2. Input flight plan for a flight including SID, STAR, and ILS runway selection
3. Went through all MCDU menu items - pulled weather TAFs, METARs, ATIS info
4. Connected to VATSIM at EGLL, viewed ATIS and METAR - then changed the setting to pull via VATSIM in the MCDU
5. Pulled and checked for accuracy

Remained on the ground on stand for the testing in this instance but all seemed good to me - really nice addition though a lot of options remained greyed out in Options but understand more to be implemented and that'll cover that

Issues (If Any): 
Overall Rating: Good
Conclusions of the testing: Love it - especially pulling METARs and ATIS info directly into the MCDU! Awesome work, thanks!

**Review by Carter633 — APPROVED** [review]
Name: British Avgeek YT -- Carter633 #1436
Date of testing: Current date 29/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1280
Object of test: Check out the new basic ATSU, AIDS, CFDS functionality

Testing technique: Test 2 - Full Flight on VATSIM
Steps you've made to test the Pull Request:

1. Spawned C&D at EGPH, switched avionics on and inserted plan into MCDU
2. Input flight plan for a flight including SID, STAR, and ILS runway selection
3. Went through all MCDU menu items - pulled weather TAFs, METARs, ATIS info
4. Fiddled with lots of ATIS info for active VATSIM airfields, and pulled METARs too for Departure airport and arrival airport, and a few en-route too
5. Kept messing about with it en-route, and used the data to set up the arrival perf page before getting the ATIS on VATSIM - all added up perfectly!

Flight Notes: Absolutely LOVE this! I was even able to pre-empt a runway change at Edinburgh when the new ground controller came online as I could see the wind had switched to a significant tailwind with the latest METAR... sure enough 5mins later a runway change happened.


Issues (If Any):
Overall Rating: Good
Conclusions of the testing: Works really well - as Borat would say - I LIKE! VERY NICE!

**Comment by Carter633** [comment]
Pulling loads of en-route ATIS info but also this is my destination ATIS - before I can even see it on VATSIM vPilot as I'm not quite in range yet! Really great
Microsoft Flight Simulator 29_10_2020 13_39_52 (2)

**Comment by mathis650** [comment]
Love it! Works great. Can't wait for this to be merged!!

**Comment by danbw9** [comment]
Name: danbw9#7028
Date of testing: Current date 29/10/2020
Version of the sim: 1.9.5.0
PR Tested: #1280
Object of test: Check out the new basic ATSU, AIDS, CFDS functionality

Testing technique: Tested in two different scenarios and check any discrepancies or bugs. Testing more ATSU functions in this comment.

Steps you've made to test the Pull Request:

Spawn on RWY at UAAA. Check all the system tests work as they should - those that are currently functional. All fine and no significant drop in FPS on my side as was previously reported. Selection of AIDS page works. The delay in page selection change with the 'SEL' in blue is spot on. Requesting METARs was realistic with the waiting time. Tried grabbing METARs from all around the world, requesting 1 and 3 airports at a time. 

Checked ability to clear any received messages with the CLR key on MCDU

Checked correct NAV database (using Navigraph latest version) and correct engine type.

Tried out Free Text to work sending MSGes to myself and to others and worked all fine. 

Spawn C&D for flight from UAAA to OMDB - normal setup, start up, taxi, take off and CRZ. Normal flight. 

Checked out AIDS and CFDS pages and they are all fine. Didn't do any systems tests in flight but will report in later PR test report or as the PR advances

METARs and TAFs are all accurate and up to date like other sources I tried. The delay of 7 mins is accurate like the real 320.

Flight Notes: 

Startup: Everything normal

Taxi/Takeoff : N/A

Cruise: Checking various. Meteoblue (default MSFS source) cannot pick up some METARs/TAFs. Tried grabbing FSIA and FHAW for example and had to change my source to VATSIM to get them. D-ATIS for OMDB works all fine. Bug described below though.

Descent/Landing: Checking and requesting multiple D-ATIS'es and METARs for DEST and it didnt get overloaded

Shutdown: N/A

Issues (If Any): small bug. See final paragraph of this comment along with the screenshots at bottom.
Overall Rating: Good
Conclusions of the testing: Works really well.  I will continue to test this in the next few days and try out all the features in the OPTIONS menu. At first glance they work all fine (AIDS and CFDS). Thanks @wpine215 for your work. Super useful, ties in well with that the METAR data in the new update is supposedly working better now...after all those bugs are fixed. 

When requesting ATIS via AOC (my source being VATSIM), your DEST (eg. OMDB) is in a smaller font (1st screenshot) as shown below. Once requested and you view message it's all blank (2nd screenshot)
Screenshot (488)
Screenshot (489)

Once I manually input OMDB, larger font is shown (nice detail!) and once requested, the received message works as it should.
Screenshot (491)

Screenshot (490)

**Comment by mathis650** [comment]
Any way to get this PR working on the latest version of the game? When I try it right now I'm getting the black MCDUs

**Comment by wpine215** [comment]
PR converted back to draft until new changes are implemented to make it compatible with the latest update. Also awaiting changes to backend API which would require endpoints and calls to be updated.

**Comment by Carter633** [comment]
Still reverted to draft or back ready to test as the label is still on post patch? Tested it the other day so if it's ready I'm happy to run more tests if it's ready?

**Comment by nistei** [comment]
I am working on porting this to the new API

**Comment by ntlgr** [comment]
>  [OPTIONS] Implemented option for ADIRS alignment time

Team, how does that work? is it under a menu or the trick with the IRS3 knob?

**Comment by lukecologne** [comment]
Menu

**Comment by VCC18DT-Hornet** [comment]
I gather that this PR still doesn't work with MSFS 1.10.8

**Comment by myselfaaryan** [comment]
> I gather that this PR still doesn't work with MSFS 1.10.8
> 

Yes because #1761 hasn't been merged yet.

**Comment by VCC18DT-Hornet** [comment]
I will give it another test

**Comment by VCC18DT-Hornet** [comment]
it all checks out everything seems to be working as it should.

**Comment by wpine215** [comment]
Needs extensive testing of the telex (free text) system - apparently some users have issues with the telex update loop not running (therefore not sending pings to the server or fetching incoming messages).

**Pull request merged**